Ensure isa object method still works after using isa infix operator

FractalBoy opened the pull request:
Fixes #18679 

Added some tests to ensure isa object method still works after using isa infix operator.

All isa tests pass.

FractalBoy committed:
Ensure isa object method still works after using isa infix operator

FractalBoy committed:
Add Marc Reisner to AUTHORS file

Leont commented on t/op/isa.t:
```diff
@@ -11,7 +11,7 @@ use strict;
 use feature 'isa';
 no warnings 'experimental::isa';
 
-plan 11;
```
Can you please give your commits more descriptive titles? «Fix github issue 18679» doesn't tell a casual reader what it happening here, «Ensure isa object method still works after using isa infix operator» does (though mentioning github issue 18679 in the description would still be useful)

Leont commented on AUTHORS:
```diff
@@ -800,6 +800,7 @@ Manuel Valente                 <mvalente@idealx.com>
 Marc Green                     <marcgreen@cpan.org>
 Marc Lehmann                   <pcg@goof.com>
 Marc Paquette                  <Marc.Paquette@Softimage.COM>
+Marc Reisner                   <reisner.marc@gmail.com>
```
This commit would be better described as «Add Marc Reisner to AUTHORS file»

FractalBoy commented:
@Leont addressed your comments

leonerd commented on universal.c:
```diff
@@ -216,12 +216,7 @@ Perl_sv_isa_sv(pTHX_ SV *sv, SV *namesv)
 
-    /* This abuse of gv_fetchmeth_pv() with level = 1 skips the UNIVERSAL
-     * lookup
-     * TODO: Consider if we want a NOUNIVERSAL flag for requesting this in a
```
Unless the `-1` part here really is relating to that NOUNIVERSAL behaviour, please leave this TODO comment in. It's a reminder to me that we still want that behaviour.

FractalBoy replied:
I thought the entire comment was one idea - I'll put back the TODO part of it.

FractalBoy replied:
Done - in commit ec8dc484c1

leonerd commented:
The new tests seem a good thing to check.

Before I approve it I would like to check that the new code still correctly skips a call to the actual `UNIVERSAL::isa`, as my previous code did for purely efficiency reasons. There's no point doing a method lookup on the `->isa` method only to find it's the basic one from UNIVERSAL, when the `else` side of the `if` condition already implements the same behaviour, just more efficiently.

FractalBoy commented:
>Before I approve it I would like to check that the new code still correctly skips a call to the actual `UNIVERSAL::isa`, as my previous code did for purely efficiency reasons. There's no point doing a method lookup on the `->isa` method only to find it's the basic one from UNIVERSAL, when the `else` side of the `if` condition already implements the same behaviour, just more efficiently.

@leonerd which `if`/`else` are you referring to? I don't understand how you can skip the call to `UNIVERSAL::isa` and still get the functionality from it? Why should infix `isa` function differently from `->isa` in this regard?

FractalBoy commented:
Ah, I see `sv_derived_from_sv` should be called instead of falling back to `XS_UNIVERSAL_isa`.

So perhaps the fix is not to change the `level` parameter but instead allow it (somehow) to fall through to that call.

FractalBoy committed:
Add GV_NOUNIVERSAL flag to skip UNIVERSAL lookup

For the `isa` infix operator, we can fall back to `sv_derived_from_sv`
instead of looking up UNIVERSAL::isa. Passing GV_NOUNIVERSAL will tell
gv_fetchmeth_internal not to look at UNIVERSAL for the method, and instead
return NULL. Then `Perl_sv_isa_sv` will skip the if block and check
`sv_derived_from_sv` which does the same thing UNIVERSAL::isa would have done.

leonerd commented:
Ahah, a new `GV_NOUNIVERSAL` flag. That sounds like a proper fix for my TODO comment.

Having added the flag though I wonder if it needs extra documentation and so on now...

FractalBoy committed:
Add documentation for new GV_NOUNIVERSAL flag

FractalBoy commented:
@leonerd Documentation updated

leonerd commented:
That looks good now. Waiting to see what the CI tests make of it...

leonerd commented:
LGTM

Cygwin smoker is upset but then it's always upset, so I've taken to ignoring it.

Thoughts @Leont ?

Leont commented:
LGTM
